airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/pipeline.py

```diff
@@ -0,0 +1,209 @@
+            version=current_metadata["dockerImageTag"],
+            build_docker_image=build_docker_image,
+            ci_report_bucket=connector_context.ci_report_bucket,
+            report_output_prefix=connector_context.report_output_prefix,
+            is_local=connector_context.is_local,
+            git_branch=connector_context.git_branch,
+            git_revision=connector_context.git_revision,
+            gha_workflow_run_url=connector_context.gha_workflow_run_url,
+            dagger_logs_url=connector_context.dagger_logs_url,
+            pipeline_start_timestamp=connector_context.pipeline_start_timestamp,
+            ci_context=connector_context.ci_context,
+            ci_gcs_credentials=connector_context.ci_gcs_credentials,
+        )
+        pypi_context.dagger_client = connector_context.dagger_client
+        return pypi_context
+
+
+class PublishToPyPI(Step):
+    context: PyPIPublishContext
+    title = "Publish package to PyPI"
+
+    async def _run(self) -> StepResult:
+        dir_to_publish = await self.context.get_repo_dir(self.context.package_path)
+
+        files = await dir_to_publish.entries()
+        is_poetry_package = "pyproject.toml" in files
+        is_pip_package = "setup.py" in files
+
+        # Try to infer package name and version from the pyproject.toml file. If it is not present, we need to have the package name and version set
+        # Setup.py packages need to set package name and version as parameter
+        if not self.context.package_name or not self.context.version:
+            if not is_poetry_package:
+                return self.skip(
+                    "Connector does not have a pyproject.toml file and version and package name is not set otherwise, skipping."
+                )
+
+            pyproject_toml = dir_to_publish.file("pyproject.toml")
+            pyproject_toml_content = await pyproject_toml.contents()
+            contents = tomli.loads(pyproject_toml_content)
+            if (
+                "tool" not in contents
+                or "poetry" not in contents["tool"]
+                or "name" not in contents["tool"]["poetry"]
+                or "version" not in contents["tool"]["poetry"]
+            ):
+                return self.skip(
+                    "Connector does not have a pyproject.toml file which specifies package name and version and they are not set otherwise, skipping."
+                )
+
+            self.context.package_name = contents["tool"]["poetry"]["name"]
+            self.context.version = contents["tool"]["poetry"]["version"]
+
+        print(
+            f"Uploading package {self.context.package_name} version {self.context.version} to {'testpypi' if self.context.test_pypi else 'pypi'}..."
+        )
+
+        if is_pip_package:
+            # legacy publish logic
+            pypi_username = self.context.dagger_client.set_secret("pypi_username", "__token__")
```

Comment:
💡Given that we are following a simple convention of
1. Choose docker image
2. Set secrets
3. Mount paths
4. Run command

You may want to look at `SimpleDockerStep`. Heres an example of it in use

https://github.com/airbytehq/airbyte/blob/99488ddcc50fd16b8753019864e80b37b947a76d/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/metadata/pipeline.py#L24

Reply:
I tried to convert it this way but due to the branching and some extras like writing certain files it got messy and IMHO it's clearer when cleaned up like @alafanechere described above - I'm still working through some comments, let me finish that and have a look at the final results, happy to do another pass.